### PR TITLE
Setup backend endpoint to send JSON data

### PR DESCRIPTION
**Pull request by Naomi-Dennis**

### CODE REVIEW

File: lib/mind_the_gapp_web/router.ex [hunk 1]
```diff
@@ -11,12 +11,13 @@ defmodule MindTheGappWeb.Router do
 
   pipeline :api do
     plug :accepts, ["json"]
+    plug CORSPlug, origin: "http://localhost:8000"
```
Comment (damonkelley) [inline lib/mind_the_gapp_web/router.ex]: I have a feeling this should not be hardcoded to `localhost`. This will break when it is deployed, yeah?

File: test/mind_the_gapp/arrival_time_test.exs [hunk 2]
```diff
@@ -0,0 +1,75 @@
+defmodule MindTheGapp.ArrivalTimeTest do
+  use MindTheGapp.DataCase, async: true
+  alias MindTheGapp.Mta.ArrivalTime
+
+  setup_all :create_fake_arrival_time_data
+
+  def create_fake_arrival_time_data(_context) do
+    clear_database()
```
Comment (damonkelley) [inline test/mind_the_gapp/arrival_time_test.exs]: This is usually taken care of by the `DataCase`. It should be opening a transaction and then rolling it back for every test. Can we check on that?

File: test/mind_the_gapp/arrival_time_test.exs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    test "returns current day to seventh day before" do
+      query_results = ArrivalTime.difference_in_arrival_times
+      current_day = query_results |> hd |> hd
+      last_day = query_results |> Enum.take(-1) |> hd |> hd
+      number_of_days_retrieved = (DateTime.diff(current_day, last_day) / (3600 * 24)) |> Kernel.trunc
+      assert number_of_days_retrieved == 7
+    end
+
+    test "returns at most eight rows, if more than eight rows are available" do
+      {:ok, db} = query_all_rows()
```
Comment (damonkelley) [inline test/mind_the_gapp/arrival_time_test.exs]: I think you could use ecto here. `MindTheGap.Repo.all(ArrivalTime)`

File: test/mind_the_gapp_web/controllers/arrival_time_controller_test.exs [hunk 4]
```diff
@@ -0,0 +1,68 @@
+defmodule MindTheGappWeb.ArrivalTimeControllerTest do
+  use MindTheGappWeb.ConnCase, async: true
+  use Timex
+
+  setup_all :create_fake_arrival_time_data
+
+  def create_fake_arrival_time_data(_context) do
+    clear_database()
```
Comment (damonkelley) [inline test/mind_the_gapp_web/controllers/arrival_time_controller_test.exs]: Same comment here. Tests should be running in transactions so no need to manually clean up the data.

File: test/mind_the_gapp_web/controllers/arrival_time_controller_test.exs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+      conn = get(conn, "/")
+      {:ok, retrieved_objects} = conn.resp_body
+                                 |> Jason.decode
+      retrieved_keys = retrieved_objects
+                       |> List.first
+                       |> Map.keys
+      assert retrieved_keys == expected_attributes
+    end
+
+    test "GET / returns a list of JSON objects, the 'meanArrivalDifferences' attributes are floats", %{conn: conn} do
```
Comment (damonkelley) [inline test/mind_the_gapp_web/controllers/arrival_time_controller_test.exs:46]: Do we care about this? I wonder if the query test is sufficient?

For these types of controller/integration tests I like to keep the tests focused on what the controller might do.
* Does the data match the *shape* I expect
* What happens when give it bad data/input (for POST request)
* How does it respond to edge cases (i.e. no data for the last 7 days).

File: test/mind_the_gapp_web/controllers/arrival_time_controller_test.exs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+     %MindTheGapp.ArrivalTime{estimated: date_time, actual: actual_time}
+     |> MindTheGapp.Repo.insert!()
+
+    date_time |> DateTime.add(3600 * -6)
+  end
+  describe "Arrival Time Controller" do
+    test "GET / returns a JSON object", %{conn: conn} do
+      conn = get(conn, "/")
+      assert json_response(conn, 200)
+    end
```
Comment (damonkelley) [inline test/mind_the_gapp_web/controllers/arrival_time_controller_test.exs:33]: I might bring in the next and last test here as well. Essentially this is the happy path test and we can assert that we got a 200 and the data matches what we expect.
```suggestion
    test "GET / returns the mean arrival", %{conn: conn} do
      response = get(conn, "/") |> json_responsee(200)
      assert [
        %{day: "Today", meanArrivalDifference: _average},
        %{day: "Yesterday", meanArrivalDifference: _average}
        | _rest
      ] = response
    end
```

File: test/mind_the_gapp/arrival_time_test.exs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+     |> MindTheGapp.Repo.insert!()
+
+    date_time |> DateTime.add(3600 * -6)
+  end
+  
+  describe "ArrivalTime" do
+    test "returns current day to seventh day before" do
+      query_results = ArrivalTime.difference_in_arrival_times
+      current_day = query_results |> hd |> hd
+      last_day = query_results |> Enum.take(-1) |> hd |> hd
```
Comment (damonkelley) [inline test/mind_the_gapp/arrival_time_test.exs:31]: It would be more idiomatic to use pattern matching instead of calling `hd` multiple times.

```suggestion
      [current_day, _average] = List.first(result)
      [last_day, _average] = List.last(result)
```

File: test/mind_the_gapp/arrival_time_test.exs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+  def generate_fake_arrival_data() do
+    starting_date = DateTime.utc_now
+
+     1..50
+     |> Enum.reduce(starting_date, &persist_arrival_time/2)
+  end
+
+  def persist_arrival_time(_x, date_time) do
+    actual_time = DateTime.add(date_time, Enum.random(-10..10) * 60)
```
Comment (damonkelley) [inline test/mind_the_gapp/arrival_time_test.exs:20]: Do we have a test to ensure the query is taking the average of the arrival times?

File: test/mind_the_gapp_web/views/arrival_time_view_test.exs [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        %{day: "Today", meanArrivalDifferences: 0.1},
+        %{day: "Yesterday", meanArrivalDifferences: 1.5},
+        %{day: "2 Days Ago", meanArrivalDifferences: 0},
+        %{day: "3 Days Ago", meanArrivalDifferences: 1.0},
+        %{day: "4 Days Ago", meanArrivalDifferences: -1.0},
+        %{day: "5 Days Ago", meanArrivalDifferences: -0.4},
+        %{day: "6 Days Ago", meanArrivalDifferences: 4},
+      ]
+
+      #assert expected_payload == ArrivalTimeView.render("index.json", data: arrival_averages)
```
Comment (damonkelley) [inline test/mind_the_gapp_web/views/arrival_time_view_test.exs]: Can this be removed?

File: test/test_helper.exs [hunk 10]
```diff
@@ -1,2 +1 @@
 ExUnit.start()
-Ecto.Adapters.SQL.Sandbox.mode(MindTheGapp.Repo, :manual)
```
Comment (damonkelley) [inline test/test_helper.exs:2]: Is there a reason this was removed? IIRC this is needed to get proper database support in test.

File: lib/mind_the_gapp_web/views/arrival_time_view.ex [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+    days_ago = calculate_difference_between_days(today, day)
+    day_label = ""
+    day_label = case days_ago do
+      0 ->
+        "Today"
+      1 ->
+        "Yesterday"
+      _ ->
+        "#{days_ago} Days Ago"
+    end
```
Comment (damonkelley) [inline lib/mind_the_gapp_web/views/arrival_time_view.ex]: Does it makes sense to extract this into a function with an intention revealing name?

File: .travis.yml [hunk 12]
```diff
@@ -12,15 +12,19 @@ jobs:
   include:
     - name: "Test"
       stage: test
-      env:
-        - MIX_ENV=test
       script:
         - mix deps.get
-        - mix ecto.create
-        - mix ecto.migrate
+        - mix deps.update db_connection
```
Comment (damonkelley) [inline .travis.yml]: what does the `db_connection` do?

**Review by damonkelley — CHANGES_REQUESTED** [review]

File: lib/mind_the_gapp_web/router.ex (hunk 1 above)
Comment (Naomi-Dennis) [reply lib/mind_the_gapp_web/router.ex]: I'll refactor it to be an env variable

File: test/mind_the_gapp/arrival_time_test.exs (hunk 2 above)
Comment (Naomi-Dennis) [reply test/mind_the_gapp/arrival_time_test.exs]: There's an error with testing migrations/schemas in Phoenix/Ecto and I think it affected the DB being cleared _every time_ before the tests are run. In 2018, Ecto was updated and the team changed how sandbox interacts with migrations/schemas. This goes for the other comments concerning `DataCase` and the other comment about running `clear_database()` in `ArrivalTimeController`

In summary, I was getting `DBConnection.OwnershipError` errors. The thread that lead me to do this and eventually remove `Ecto.Adapters.SQL.Sandbox.mode(MindTheGapp.Repo, :manual)` and add `mix deps.update db_connection` to travis is here: [Issue thread](https://github.com/elixir-ecto/ecto/issues/2782)

File: test/mind_the_gapp_web/controllers/arrival_time_controller_test.exs (hunk 5 above)
Comment (Naomi-Dennis) [reply test/mind_the_gapp_web/controllers/arrival_time_controller_test.exs:46]: I thought it would test the shape of the data.

File: test/mind_the_gapp/arrival_time_test.exs (hunk 8 above)
Comment (Naomi-Dennis) [reply test/mind_the_gapp/arrival_time_test.exs:20]: We do not; I will add that

File: test/mind_the_gapp_web/views/arrival_time_view_test.exs (hunk 9 above)
Comment (Naomi-Dennis) [reply test/mind_the_gapp_web/views/arrival_time_view_test.exs]: Yup

File: lib/mind_the_gapp_web/views/arrival_time_view.ex (hunk 11 above)
Comment (Naomi-Dennis) [reply lib/mind_the_gapp_web/views/arrival_time_view.ex]: I think it could go either way. Do you think it would be more understandable if it was extracted to a function?

File: test/mind_the_gapp/arrival_time_test.exs (hunk 2 above)
Comment (damonkelley) [reply test/mind_the_gapp/arrival_time_test.exs]: Oh i see. Thanks for that context.

I poked around at this locally, and I think I understand why this is happening. The `DataCase` will checkout a connection during the `setup` phase, but I think you saw this error because you are seeding data during the `setup_all` phase. If you change the `setup_all` to `setup` and add this like back into `test/test_helper.exs`, it should work.

```
Ecto.Adapters.SQL.Sandbox.mode(MindTheGapp.Repo, :manual)
```

My guess is that `setup_all` is called before any `setup` callbacks are run. You were getting ownership errors because the process your test was running in had not yet checkout a connection from the sandbox. 

In practice, I very rarely use a `setup_all` or `after_all` hook in my tests.

File: lib/mind_the_gapp_web/views/arrival_time_view.ex (hunk 11 above)
Comment (damonkelley) [reply lib/mind_the_gapp_web/views/arrival_time_view.ex]: I prefer to have smaller functions. I think a good litmus test is SRP. Does this function have a single reason to change?

File: test/mind_the_gapp/arrival_time_test.exs (hunk 2 above)
Comment (damonkelley) [reply test/mind_the_gapp/arrival_time_test.exs]: I did this locally and got all of the tests to pass without that error.

File: test/test_helper.exs (hunk 10 above)
Comment (damonkelley) [reply test/test_helper.exs:2]: FYI, it looks like this didn't make it back in? That probably means that the tests are continuously adding data without it getting cleaned up.
